### PR TITLE
Set OTel schema url

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Contributes to https://github.com/Azure/azure-sdk-for-net/issues/31650

### CODE REVIEW

File: sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs [hunk 1]
```diff
@@ -329,6 +338,10 @@ public void CanSetActivitySourceAndDiagnosticSourceActivitiesTogether()
             CollectionAssert.Contains(diagnosticSourceActivity.Tags, new KeyValuePair<string, string>("Attribute2", "2"));
             CollectionAssert.Contains(diagnosticSourceActivity.Tags, new KeyValuePair<string, string>("Attribute3", "3"));
 
+            // Since both ActivitySource and DiagnosticSource listeners are used, we should see the az.schema_url tag set even in diagnostic source because they use the same
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs:341]: I think this is expected.. @lmolkova please keep me honest

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 2]
```diff
@@ -292,8 +292,11 @@ public void AddLink(string traceparent, string tracestate, IDictionary<string, s
             public Activity? Start()
             {
                 _currentActivity = StartActivitySourceActivity();
-
-                if (_currentActivity == null)
+                if (_currentActivity != null)
+                {
+                    _currentActivity.AddTag("az.schema_url", "https://opentelemetry.io/schemas/1.17.0");
```
Comment (lmolkova) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:297]: I suggest bundling this change with mappings and actually complying to otel schema 1.17.0, (we're not compliant now)

Changes needed (https://github.com/Azure/azure-sdk/blob/main/docs/tracing/distributed-tracing-conventions.yml)


HTTP:
- `serviceRequestId` -> `az.service_request_id`
- `clientRequestId` -> `az.client_request_id`
- add `net.peer.name` (endpoint)
- add `net.peer.port` (if not 443), i.e. probably never, but it should be of int type
- http.status_code should be int

Messaging:
- add `messaging.system` - either 'eventhubs' or 'servicebus'
- add [`messaging.operation`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#operation-names) (one of `publish`, `receive` or `process`)
- `message_bus.destination` -> `messaging.destination.name`
- `peer.address` -> `net.peer.name`

Or, if you prefer to keep PRs small and would be sending others with attribute changes before next core release, we can do this as well :)

File: sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs (hunk 1 above)
Comment (lmolkova) [reply sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs:341]: yep, if both are used, there is nothing we can do to keep DiagSource path unchanged. But it's not a common (or valid) scenario to have both.

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs (hunk 2 above)
Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:302]: I've added the HTTP changes in this PR other than enabling suppression by default (but perhaps that is already done?). I will do the messaging changes in a separate PR.

File: sdk/core/Azure.Core/CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,8 @@
 
 ### Features Added
 
+- ActivitySource activities that are used when using the [experimental OpenTelemetry support](https://devblogs.microsoft.com/azure-sdk/introducing-experimental-opentelemetry-support-in-the-azure-sdk-for-net/) will include the `az.schema_url` tag indicating the OpenTelemetry schema version.
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/CHANGELOG.md:7]: Todo - document the breaking changes.

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 4]
```diff
@@ -232,7 +237,7 @@ public void AddTag(string name, string value)
                 }
                 else
                 {
-                    _currentActivity?.AddTag(name, value!);
+                    _currentActivity?.AddObjectTag(name, value);
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:240]: This method already existed thanks to @pakrym

File: sdk/core/Azure.Core/tests/RequestActivityPolicyTests.cs [hunk 5]
```diff
@@ -313,21 +313,31 @@ public async Task ActivitySourceActivityStartedOnRequest()
...
+
+                CollectionAssert.DoesNotContain(activity.TagObjects, new KeyValuePair<string, string>("requestId", clientRequestId));
+                CollectionAssert.Contains(activity.TagObjects, new KeyValuePair<string, string>("az.client_request_id", clientRequestId));
+
+                CollectionAssert.DoesNotContain(activity.TagObjects, new KeyValuePair<string, string>("serviceRequestId", "server request id"));
+                CollectionAssert.Contains(activity.TagObjects, new KeyValuePair<string, string>("az.service_request_id", "server request id"));
+
+                CollectionAssert.Contains(activity.TagObjects, new KeyValuePair<string, string>("net.peer.name", "example.com"));
+
+                CollectionAssert.DoesNotContain(activity.TagObjects, new KeyValuePair<string, int>("net.peer.port", 443));
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/tests/RequestActivityPolicyTests.cs:338]: Add a test where the port is not 443.

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 6]
```diff
@@ -292,8 +297,11 @@ public void AddLink(string traceparent, string tracestate, IDictionary<string, s
             public Activity? Start()
             {
                 _currentActivity = StartActivitySourceActivity();
-
-                if (_currentActivity == null)
+                if (_currentActivity != null)
+                {
+                    _currentActivity.AddTag("az.schema_url", "https://opentelemetry.io/schemas/1.17.0");
```
Comment (lmolkova) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:302]: nit: my crystal ball tells me we'll need to support another version soon, so it'd be nice to move it to some constant and update all at once along with tests :)

**Review by lmolkova — APPROVED** [review]
Thank you!

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs (hunk 2 above)
Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs]: Looks like we also need messaging.batch.message_count for batch operations in messaging.

Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs]: Potentially also message.destination.name and message.source.name?
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#producer-attributes

**Pull request merged**